File: 10kindsofpeople.py

r, c = map(int, input().split())
grid = [list(map(int, list(input()))) for _ in range(r)]
mesh_grid = [[-1 for j in range(c)] for i in range(r)]

mesh_index = 0
for i in range(r):
    for j in range(c):
        
        if mesh_grid[i][j] != -1:
            continue
        
        num = grid[i][j]
        to_check = [(i, j)]

        check_index = 0
        while check_index < len(to_check):
            
            check_i = to_check[check_index][0]
            check_j = to_check[check_index][1]
            
            mesh_grid[check_i][check_j] = mesh_index
            
            if check_i > 0:
                if grid[check_i-1][check_j] == num and mesh_grid[check_i-1][check_j] == -1:
                    to_check.append((check_i-1, check_j))
            if check_i < r-1:
                if grid[check_i+1][check_j] == num and mesh_grid[check_i+1][check_j] == -1:
                    to_check.append((check_i+1, check_j))
            if check_j > 0:
                if grid[check_i][check_j-1] == num and mesh_grid[check_i][check_j-1] == -1:
                    to_check.append((check_i, check_j-1))
            if check_j < c-1:
                if grid[check_i][check_j+1] == num and mesh_grid[check_i][check_j+1] == -1:
                    to_check.append((check_i, check_j+1))
            
            check_index += 1
        
        mesh_index += 1

n = int(input())
for _ in range(n):
    
    r1, c1, r2, c2 = map(int, input().split())
    condition = "neither"
    if mesh_grid[r1-1][c1-1] == mesh_grid[r2-1][c2-1]:
        condition = "binary" if grid[r1-1][c1-1] == 0 else "decimal"
    print(condition)

File: 10kindsofpeople.cpp

#include <iostream>
#include <vector>
#include <string>

using namespace std;

void dfs(vector<vector<bool>>& grid, vector<vector<int>>& grid_signs, int sign, int i, int j) {
    
    grid_signs[i][j] = sign;

    int h = grid.size();
    int w = grid[0].size();
    
    if (i > 0) {
        if (!grid_signs[i-1][j] && grid[i][j] == grid[i-1][j]) {
            dfs(grid, grid_signs, sign, i-1, j);
        }
    }
    if (i < h-1) {
        if (!grid_signs[i+1][j] && grid[i][j] == grid[i+1][j]) {
            dfs(grid, grid_signs, sign, i+1, j);
        }
    }
    if (j > 0) {
        if (!grid_signs[i][j-1] && grid[i][j] == grid[i][j-1]) {
            dfs(grid, grid_signs, sign, i, j-1);
        }
    }
    if (j < w-1) {
        if (!grid_signs[i][j+1] && grid[i][j] == grid[i][j+1]) {
            dfs(grid, grid_signs, sign, i, j+1);
        }
    }
    
}

int main() {
    
    int r, c;
    cin >> r >> c;
    
    vector<vector<bool>> grid(r, vector<bool>(c));
    for (int i = 0; i < r; i++) {
        string row;
        cin >> row;
        for (int j = 0; j < c; j++) {
            grid[i][j] = (row[j] == '1');
        }
    }
    
    int n;
    cin >> n;

    vector<vector<int>> grid_signs(r, vector<int>(c, 0));
    int current_sign = 1;

    for (int i = 0; i < n; i++) {
        
        int r1, c1, r2, c2;
        cin >> r1 >> c1 >> r2 >> c2;
        r1--;
        c1--;
        r2--;
        c2--;
        
        if (grid[r1][c1] == grid[r2][c2]) {
            if (grid_signs[r1][c1] == 0 && grid_signs[r2][c2] == 0) {
                dfs(grid, grid_signs, current_sign, r1, c1);
                current_sign++;
            }
            if (grid_signs[r1][c1] == grid_signs[r2][c2]) {
                if (grid[r1][c1]) {
                    cout << "decimal";
                } else {
                    cout << "binary";
                }
            } else {
                cout << "neither";
            }
        } else {
            cout << "neither";
        }

        cout << endl;

    }
    
}
